core/src/main/java/com/linecorp/armeria/internal/common/Http1ObjectEncoder.java

```diff
@@ -112,13 +116,45 @@ protected final ChannelFuture writeNonInformationalHeaders(int id, HttpObject co
         } else {
             f = write(id, converted, false);
             if (endStream) {
-                f = write(id, LastHttpContent.EMPTY_LAST_CONTENT, true);
+                f = combine(f, write(id, LastHttpContent.EMPTY_LAST_CONTENT, true));
             }
         }
         ch.flush();
         return f;
     }
 
+    private ChannelPromise combine(ChannelFuture first, ChannelFuture second) {
+        final ChannelPromise promise = channel().newPromise();
+        final FutureListener<Void> listener =
+                new FutureListener<Void>() {
+                    private final AtomicInteger counter = new AtomicInteger(2);
```

Comment:
We can just use a non-volatile `int` field because the callback will always be invoked from the event loop thread.

Reply:
```suggestion
                    private int remaining = 2;
```